tests/unit/pubsublocal.tcl

```diff
@@ -0,0 +1,227 @@
+    }
+
+    test "UNSUBSCRIBE from non-subscribed channels" {
+        set rd1 [redis_deferring_client]
+        assert_equal {0 0 0} [unsubscribelocal $rd1 {foo bar quux}]
+
+        # clean up clients
+        $rd1 close
+    }
+
+    test "PUBSUB command basics" {
+        r pubsub local numsub abc def
+    } {abc 0 def 0}
+
+    test "PUBLISHLOCAL/SUBSCRIBELOCAL with two clients" {
+        set rd1 [redis_deferring_client]
+        set rd2 [redis_deferring_client]
+
+        assert_equal {1} [subscribelocal $rd1 {chan1}]
+        assert_equal {1} [subscribelocal $rd2 {chan1}]
+        assert_equal 2 [r publishlocal chan1 hello]
+        assert_equal "chan1 2" [r pubsub local numsub chan1]
+        assert_equal "chan1" [r pubsub local channels]
+
+        # clean up clients
+        $rd1 close
+        $rd2 close
+    }
+
+    test "PUBLISHLOCAL/SUBSCRIBELOCAL with PUBLISH/SUBSCRIBE" {
+        set rd1 [redis_deferring_client]
+        set rd2 [redis_deferring_client]
+
+        assert_equal {1} [subscribelocal $rd1 {chan1}]
+        assert_equal {1} [subscribe $rd2 {chan1}]
+        assert_equal 1 [r publishlocal chan1 hello]
+        assert_equal 1 [r publish chan1 hello]
+        assert_equal "chan1 1" [r pubsub local numsub chan1]
+        assert_equal "chan1 1" [r pubsub numsub chan1]
+        assert_equal "chan1" [r pubsub local channels]
+        assert_equal "chan1" [r pubsub channels]
+    }
+}
+
+start_server {tags {"pubsublocal"}} {
+start_server {} {
+    set node_0 [srv 0 client]
+    set node_0_host [srv 0 host]
+    set node_0_port [srv 0 port]
+    set node_0_pid [srv 0 pid]
+
+    set node_1 [srv -1 client]
+    set node_1_host [srv -1 host]
+    set node_1_port [srv -1 port]
+    set node_1_pid [srv -1 pid]
+
+    test {setup replication for following tests} {
+        $node_1 replicaof $node_0_host $node_0_port
+        wait_for_sync $node_1
+    }
+
+    test {publish message to replica and receive on master} {
+        set rd0 [redis_deferring_client node_0_host node_0_port]
+        set rd1 [redis_deferring_client node_1_host node_1_port]
+
+        assert_equal {1} [subscribelocal $rd0 {chan1}]
+        $rd1 PUBLISHLOCAL chan1 hello
+        assert_equal {message chan1 hello} [$rd0 read]
+        $rd1 PUBLISHLOCAL chan1 world
+        assert_equal {message chan1 world} [$rd0 read]
+    }
+
+    test {setup reverse replication for following tests} {
+        $node_1 replicaof no one
+        $node_0 replicaof $node_1_host $node_1_port
+        wait_for_sync $node_0
+    }
+
+    test {publish message to master and receive on replica} {
+        set rd0 [redis_deferring_client node_0_host node_0_port]
+        set rd1 [redis_deferring_client node_1_host node_1_port]
+
+        assert_equal {1} [subscribelocal $rd1 {chan1}]
+        $rd0 PUBLISHLOCAL chan1 hello
+        assert_equal {message chan1 hello} [$rd1 read]
+        $rd0 PUBLISHLOCAL chan1 world
+        assert_equal {message chan1 world} [$rd1 read]
+    }
```

Comment:
I don't understand this test. Both of these tests are sending the publish to the primary and reading on the replica.

Reply:
Yeah my bad, removed the reverse replication.